### DISCUSSION TITLE
enhance coolprop with gradient data

### DISCUSSION THREAD

**Discussion by stockh0lm** [CoolProp/CoolProp#2421] [Ideas]
I did a quick hack and interpolated cool prop data with splines using their gradient feature. Now, I can use automatic differentiation (AD) to optimize heat pump cycles, which offers significant new and better options over alternatives. 

Of course, there are better solutions than splines. I would like to hear experts' opinions on how to deal with the non-differentiable phase transitions so that I can use AD continuously over the intervals. 

I can share my hacky implementation if you are interested.

**Comment by ibell** [comment]
This is a similar sort of problem to what we had with Modelica and
derivatives within the two-phase region. Did you see our paper on that
theme?

**Reply by stockh0lm** [reply]
No, I didn't. A quick search didn't bring up prominent candidate papers. Can you provide the authors, year, and title, please?

**Reply by ibell** [reply]
This is the one I had in mind : https://www.mdpi.com/1996-1073/7/3/1621

**Reply by stockh0lm** [reply]
Now I looked at the paper. It is outside my center of expertise. What it concludes is basically. 

Smoothing the Density Function:

    Best for overall accuracy: It balances robustness and accuracy well, particularly for eliminating discontinuities in the liquid saturation line. I think this method is recommended if your primary goal is high accuracy and robustness.
    Good for transient conditions: Especially suitable when handling transitions between phases (e.g., evaporation and condensation).

Mean Densities Method:

    Best for speed: It is computationally efficient and handles phase changes well without introducing large mass or energy balance errors. However, it struggles with flow reversal, so it should be avoided in systems where flow reversals occur.

Enthalpy Limiter Method:

    Best for handling flow reversals: This method is highly effective when dealing with systems prone to flow reversals. However, it introduces some artifacts, which can slightly increase the error in the energy balance.
    
I want to make coolprop more usable for optimization, and gradients are critical. 

Without any further consideration, I used SciPy's RectBivariateSpline like this:
```

h_values = []
s_values = []

for P, T in zip(P_flat, T_flat):
    try:
        h = CP.PropsSI('H', 'P', P, 'T', T, refrigerant)
        s = CP.PropsSI('S', 'P', P, 'T', T, refrigerant)
        h_values.append(h)
        s_values.append(s)
    except:
        h_values.append(np.nan)
        s_values.append(np.nan)

# Reshape data to grid shape
h_values = np.array(h_values).reshape(P_grid.shape)
s_values = np.array(s_values).reshape(P_grid.shape)

# Handle NaNs by masking invalid entries
valid_mask = ~np.isnan(h_values) & ~np.isnan(s_values)
P_valid = P_grid[valid_mask]
T_valid = T_grid[valid_mask]
h_valid = h_values[valid_mask]
s_valid = s_values[valid_mask]

# Fit bivariate splines for enthalpy and entropy
h_spline = RectBivariateSpline(P_range, T_range, np.nan_to_num(h_values).T)
s_spline = RectBivariateSpline(P_range, T_range, np.nan_to_num(s_values).T)

```

, which produces splines like this for the phase transition:

![phase_transition](https://github.com/user-attachments/assets/a5dd7c1a-0cc3-423b-aaac-7b1646101ce0)

, which leaves some things to desire.

The awesome thing is that I can optimize over the interval or determine sensitivities in algorithms involving thermodynamic properties. Both will become increasingly important as more people awaken to their advantages. 

I would be happy to contribute to another paper in that direction and help with code, providing gradients. I am unfamiliar with the coolprops codebase and would ask for direction on how to approach that.

**Reply by ibell** [reply]
Ah this is a classic Runge phenomenon problem in your figure

**Reply by stockh0lm** [reply]
How nice that it has a name!

Do you have thoughts regarding the gradients and their implementation? Are more people (besides you) interested in this, and do they have an opinion?
